File: modules/snoozepulse-audio/android/src/main/java/expo/modules/snoozepulseaudio/YamnetClassifier.kt
package expo.modules.snoozepulseaudio

import android.content.Context
import android.util.Log
import java.io.FileInputStream
import java.nio.MappedByteBuffer
import java.nio.channels.FileChannel
import org.tensorflow.lite.Interpreter
import org.tensorflow.lite.nnapi.NnApiDelegate

/**
 * YAMNet on-device inference (ADR-21).
 *
 * The public YAMNet TFLite classification model bundled at
 * `android/src/main/assets/yamnet.tflite` (ADR-28) takes a 15,600-sample float32 waveform
 * at 16 kHz mono and emits three outputs; only `scores` is consumed here. The AudioSet
 * ontology places `Snoring` at index 38 and `Snort` at index 41.
 *
 * Task 6.1 keeps this class off the capture path. `CaptureEngine` still runs the M5
 * loudness detector until ADR-23's deletion lands in Task 6.3.
 */
internal class YamnetClassifier(private val context: Context) : SnoreClassifier {
  companion object {
    private const val TAG = "YamnetClassifier"
    private const val MODEL_ASSET = "yamnet.tflite"
    /** 0.975 s of 16 kHz mono PCM — YAMNet's fixed input length. */
    const val INPUT_SAMPLES = 15_600
    private const val NUM_CLASSES = 521
    private const val CLASS_SNORING = 38
    private const val CLASS_SNORT = 41
    private const val DELEGATE_NNAPI = "NNAPI"
    private const val DELEGATE_CPU = "CPU"
    private const val WARM_SILENCE_LEVEL = 0f
  }

  private var interpreter: Interpreter? = null
  private var nnApiDelegate: NnApiDelegate? = null
  private var resolvedDelegate: String = DELEGATE_CPU
  private var scoresOutputIndex: Int = 0
  private var scoresFrames: Int = 1
  private var inputWantsBatchDim: Boolean = false
  // Reused across calls — capture path is single-threaded (`CaptureEngine.captureLoop`).
  private var outputBuffers: MutableMap<Int, Any> = mutableMapOf()
  private var batchedInputScratch: Array<FloatArray>? = null

  override fun load() {
    if (interpreter != null) {
      return
    }
    val model = loadModelFile()
    interpreter =
      try {
        val delegate = NnApiDelegate()
        val options = Interpreter.Options().setNumThreads(2).addDelegate(delegate)
        val ip = Interpreter(model, options)
        nnApiDelegate = delegate
        resolvedDelegate = DELEGATE_NNAPI
        ip
      } catch (error: Exception) {
        Log.w(TAG, "NNAPI delegate unavailable, falling back to CPU: ${error.message}")
        nnApiDelegate?.close()
        nnApiDelegate = null
        val ip = Interpreter(model, Interpreter.Options().setNumThreads(2))
        resolvedDelegate = DELEGATE_CPU
        ip
      }
    resolveTensorShapes()
    allocateOutputBuffers()
  }

  override fun warm(): String {
    load()
    // Silence input so the first inference never fires the snore path even if the caller
    // is already forwarding results anywhere.
    val silence = FloatArray(INPUT_SAMPLES) { WARM_SILENCE_LEVEL }
    classify(silence)
    Log.i(TAG, "YAMNet warmed on delegate: $resolvedDelegate")
    return resolvedDelegate
  }

  override fun classify(patch: FloatArray): Float {
    val ip = interpreter ?: throw IllegalStateException("YamnetClassifier not loaded")
    require(patch.size == INPUT_SAMPLES) {
      "YAMNet expects $INPUT_SAMPLES samples, received ${patch.size}"
    }
    val inputTensor: Any =
      if (inputWantsBatchDim) {
        val scratch = batchedInputScratch ?: Array(1) { FloatArray(INPUT_SAMPLES) }
        batchedInputScratch = scratch
        System.arraycopy(patch, 0, scratch[0], 0, INPUT_SAMPLES)
        scratch
      } else {
        patch
      }
    ip.runForMultipleInputsOutputs(arrayOf(inputTensor), outputBuffers)
    @Suppress("UNCHECKED_CAST")
    val scores = outputBuffers[scoresOutputIndex] as Array<FloatArray>
    var snoreSum = 0f
    var snortSum = 0f
    for (frame in scores) {
      snoreSum += frame[CLASS_SNORING]
      snortSum += frame[CLASS_SNORT]
    }
    val frames = scores.size.coerceAtLeast(1).toFloat()
    val combined = (snoreSum + snortSum) / frames
    return combined.coerceIn(0f, 1f)
  }

  override fun close() {
    interpreter?.close()
    interpreter = null
    nnApiDelegate?.close()
    nnApiDelegate = null
    outputBuffers.clear()
    batchedInputScratch = null
  }

  private fun resolveTensorShapes() {
    val ip = interpreter ?: return
    val inputShape = ip.getInputTensor(0).shape()
    inputWantsBatchDim =
      when {
        inputShape.size == 1 && inputShape[0] == INPUT_SAMPLES -> false
        inputShape.size == 2 && inputShape[1] == INPUT_SAMPLES -> true
        else -> {
          Log.w(TAG, "Unexpected YAMNet input shape: ${inputShape.joinToString()}")
          inputShape.size == 2
        }
      }
    for (i in 0 until ip.outputTensorCount) {
      val shape = ip.getOutputTensor(i).shape()
      if (shape.size == 2 && shape[1] == NUM_CLASSES) {
        scoresOutputIndex = i
        scoresFrames = shape[0].coerceAtLeast(1)
        return
      }
    }
    Log.w(TAG, "Could not locate a [N, $NUM_CLASSES] output; defaulting to index 0.")
  }

  private fun allocateOutputBuffers() {
    val ip = interpreter ?: return
    outputBuffers.clear()
    for (i in 0 until ip.outputTensorCount) {
      val shape = ip.getOutputTensor(i).shape()
      outputBuffers[i] = allocateFloatBufferForShape(shape)
    }
  }

  private fun allocateFloatBufferForShape(shape: IntArray): Any =
    when (shape.size) {
      1 -> FloatArray(shape[0].coerceAtLeast(1))
      2 -> Array(shape[0].coerceAtLeast(1)) { FloatArray(shape[1].coerceAtLeast(1)) }
      3 -> Array(shape[0].coerceAtLeast(1)) {
        Array(shape[1].coerceAtLeast(1)) { FloatArray(shape[2].coerceAtLeast(1)) }
      }
      else -> {
        // YAMNet only produces 1D–3D outputs; keep a defensive fallback anyway.
        Log.w(TAG, "Unexpected output rank ${shape.size}, allocating flat float array.")
        FloatArray(shape.fold(1) { acc, dim -> acc * dim.coerceAtLeast(1) })
      }
    }

  private fun loadModelFile(): MappedByteBuffer {
    val fd = context.assets.openFd(MODEL_ASSET)
    FileInputStream(fd.fileDescriptor).use { stream ->
      return stream.channel.map(
        FileChannel.MapMode.READ_ONLY,
        fd.startOffset,
        fd.declaredLength,
      )
    }
  }
}

File: modules/snoozepulse-audio/ios/YamnetClassifier.swift
import Foundation
import TensorFlowLite

/**
 * YAMNet on-device inference (ADR-21).
 *
 * The public YAMNet TFLite classification model bundled at
 * `ios/Resources/yamnet.tflite` (ADR-28) takes a 15,600-sample float32 waveform at 16 kHz
 * mono and emits three outputs; only `scores` is consumed here. The AudioSet ontology
 * places `Snoring` at index 38 and `Snort` at index 41.
 *
 * Task 6.1 keeps this class off the capture path. `CaptureEngine` still runs the M5
 * loudness detector until ADR-23's deletion lands in Task 6.3.
 */
final class YamnetClassifier: SnoreClassifier {
  static let inputSamples = 15_600
  private static let numClasses = 521
  private static let classSnoring = 38
  private static let classSnort = 41
  private static let modelResource = "yamnet"
  private static let modelExtension = "tflite"
  private static let delegateCoreML = "CoreML"
  private static let delegateCPU = "CPU"

  enum ClassifierError: Error {
    case modelNotFound
    case notLoaded
    case inputSizeMismatch(expected: Int, actual: Int)
  }

  private var interpreter: Interpreter?
  private var scoresOutputIndex: Int = 0
  private var scoresFrames: Int = 1
  private var resolvedDelegate: String = YamnetClassifier.delegateCPU

  func load() throws {
    if interpreter != nil { return }
    guard
      let modelPath = Bundle.main.path(
        forResource: Self.modelResource,
        ofType: Self.modelExtension
      )
    else {
      throw ClassifierError.modelNotFound
    }
    let options = Interpreter.Options()
    options.threadCount = 2

    // Prefer Core ML on capable hardware; fall through to CPU on any failure.
    var delegates: [Delegate] = []
    if let coreML = CoreMLDelegate() {
      delegates.append(coreML)
      resolvedDelegate = Self.delegateCoreML
    } else {
      resolvedDelegate = Self.delegateCPU
    }

    do {
      let ip = try Interpreter(
        modelPath: modelPath,
        options: options,
        delegates: delegates.isEmpty ? nil : delegates
      )
      try ip.allocateTensors()
      interpreter = ip
    } catch {
      // Retry without the delegate. The Core ML build sometimes fails on older devices
      // even when the delegate instance itself constructed successfully.
      let ip = try Interpreter(modelPath: modelPath, options: options)
      try ip.allocateTensors()
      interpreter = ip
      resolvedDelegate = Self.delegateCPU
    }

    resolveTensorShapes()
  }

  @discardableResult
  func warm() throws -> String {
    try load()
    let silence = [Float](repeating: 0, count: Self.inputSamples)
    _ = try classify(patch: silence)
    NSLog("YAMNet warmed on delegate: %@", resolvedDelegate)
    return resolvedDelegate
  }

  func classify(patch: [Float]) throws -> SnoreClassification {
    guard let ip = interpreter else { throw ClassifierError.notLoaded }
    guard patch.count == Self.inputSamples else {
      throw ClassifierError.inputSizeMismatch(
        expected: Self.inputSamples,
        actual: patch.count
      )
    }
    let inputData = patch.withUnsafeBufferPointer { pointer -> Data in
      Data(buffer: pointer)
    }
    try ip.copy(inputData, toInputAt: 0)
    try ip.invoke()

    let output = try ip.output(at: scoresOutputIndex)
    let floatCount = output.data.count / MemoryLayout<Float>.size
    let scores = output.data.withUnsafeBytes { raw -> [Float] in
      let buffer = raw.bindMemory(to: Float.self)
      return Array(UnsafeBufferPointer(start: buffer.baseAddress, count: floatCount))
    }

    let framesInOutput = max(1, scoresFrames)
    var snoreSum: Float = 0
    var snortSum: Float = 0
    for frame in 0..<framesInOutput {
      let base = frame * Self.numClasses
      snoreSum += scores[base + Self.classSnoring]
      snortSum += scores[base + Self.classSnort]
    }
    let framesF = Float(framesInOutput)
    let snoring = min(1.0, max(0.0, snoreSum / framesF))
    let snort = min(1.0, max(0.0, snortSum / framesF))
    let combined = min(1.0, max(0.0, snoring + snort))
    return SnoreClassification(combined: combined, snoring: snoring, snort: snort)
  }

  func close() {
    interpreter = nil
  }

  private func resolveTensorShapes() {
    guard let ip = interpreter else { return }
    let outputCount = ip.outputTensorCount
    for i in 0..<outputCount {
      guard let tensor = try? ip.output(at: i) else { continue }
      let dims = tensor.shape.dimensions
      if dims.count == 2, dims[1] == Self.numClasses {
        scoresOutputIndex = i
        scoresFrames = max(1, dims[0])
        return
      }
    }
    NSLog(
      "YAMNet: could not locate a [N, %d] output tensor; defaulting to index 0.",
      Self.numClasses
    )
    scoresOutputIndex = 0
    scoresFrames = 1
  }
}
